library/lua/profiler.lua

```diff
@@ -0,0 +1,651 @@
+  local prevented = self.prevented_functions
+  local depth = 2
+  local caller = debug.getinfo(depth,'fl')
+  local child
+  if not caller then
+    return
+  end
+  local cf = caller.func
+  if cf then
+    child = get_stats(rawstats, prevented, cf, depth)
+    if child then
+      child.count = child.count + 1
+      child.time = child.time + timetaken
+      local line = caller.currentline
+      child.currentline[line] = (child.currentline[line] or 0) + 1
+    else
+      cf = DEFAULT_FILTERED_FUNC
+    end
+  else
+    cf = DEFAULT_MISSING_FUNC
+  end
+  depth = 3
+  local caller = debug.getinfo(depth,'f')
+  while caller do
+    if caller.func then
+      local this = get_stats(rawstats, prevented, caller.func, depth)
+      if this then
+        this.time = this.time + timetaken
+        if not child or child.anon then
+          this.anon_child_time = this.anon_child_time + timetaken
+        else
+          this.name_child_time = this.name_child_time + timetaken
+        end
+        local ch = this.children[cf]
+        if ch then
+          this.children[cf] = ch + 1
+          this.children_time[cf] = this.children_time[cf] + timetaken
+        else
+          this.children[cf] = 1
+          this.children_time[cf] = timetaken
+        end
+        cf = caller.func
+      else
+        cf = DEFAULT_FILTERED_FUNC
+      end
+      child = this
+    else
+      cf = DEFAULT_MISSING_FUNC
+      child = nil
+    end
+    depth = depth + 1
+    caller = debug.getinfo(depth, 'f')
+  end
+  self.lastclock = os.clock()
+end
+
+
+--
+-- This function starts the profiler.  It will do nothing
+-- if this (or any other) profiler is already running.
+--
+function _profiler.start(self)
+  if _profiler.running then
+    return
+  end
+  -- Start the profiler. This begins by setting up internal profiler state
+  _profiler.running = self
+  assert(_profiler.running)
+  self.rawstats = {}
+  self.stack = {}
+  self.depth = 0
+  if self.variant == "time" then
+    self.lastclock = os.clock()
+    debug.sethook( _profiler_hook_wrapper_by_time, "", self.sampledelay )
+  elseif self.variant == "call" then
+    debug.sethook( _profiler_hook_wrapper_by_call, "cr" )
+  else
+    print("Profiler method must be 'time' or 'call'.")
+    sys.exit(1)
+  end
+end
+
+--
+-- This writes a profile report to the output file object.  If
+-- sort_by_total_time is nil or false the output is sorted by
+-- the function time minus the time in it's children.
+--
+function _profiler.report( self, outfile, sort_by_total_time )
+
+  outfile:write
+    [[Lua Profile output created by profiler.lua. Copyright Pepperfish 2002+
+
+]]
+
+  -- This is pretty awful.
+  local terms = {}
+  if self.variant == "time" then
+    terms.capitalized = "Sample"
+    terms.single = "sample"
+    terms.pastverb = "sampled"
+  elseif self.variant == "call" then
+    terms.capitalized = "Call"
+    terms.single = "call"
+    terms.pastverb = "called"
+  else
+    assert(false)
```

Comment:
Might as well replace with `error()`, although I think the above error() should be enough.